src/main/scala/cognite/spark/v1/FlexibleDataModelCorePropertySyncRelation.scala

```diff
@@ -244,11 +247,22 @@ private[spark] class FlexibleDataModelCorePropertySyncRelation(
       fetchData(cursors).map { sr =>
         val items = sr.items
         val nextCursor = sr.nextCursor
-        val shouldStopEarly = items.isEmpty || (applyTimestampTermination && items.last.lastUpdatedTime >=
-          terminationTimeStamp)
+        val shouldStopEarly = items.nonEmpty && applyTimestampTermination && items.last.lastUpdatedTime >= terminationTimeStamp
         val next =
-          (nextCursor, shouldStopEarly) match {
-            case (Some(cursor), false) =>
+          (nextCursor, items.isEmpty, shouldStopEarly) match {
```

Comment:
I think we miss a case for (Some(cursor), false, true) ?

Reply:
Yeah, for early /sync termination we go to `_` case that also returns empty stream like the last page but callbackUrl call is currently missing there :100:

Reply:
So, this was actually on purpose, but might still be wrong? When we hit the shouldStopEarly, we will *not* save the items out. When this happens, we also do not want to save out the cursor, as we want the next run to actually use the previous cursor to get to this again.

Else we would lose the last page, wouldn't we?

Reply:
Ah, this takes care of this by only sending the sync cursor, not the query cursor, so then we're fine. It should save the last cursor, although it will have persisted that already if there was any data. If there was no data to backfill, then it would not work, and it would always trigger backfill.

Reply:
And since "nextCursor" is the cursor of the page we actually emit right now (project later on), this is exactly the next page we want to continue on.

So, yes, adding the last variation is necessary. It also fixes the always-trigger-backfill issue.

Reply:
With `shouldStopEarly` we try to save current page items if any.

I think we need to clarify when is the callback expected to fire
- always fire at the end of stream just for good measure
- only fire if the stream is empty
- some other combination of conditions

Reply:
I've redone this slight now.
The thing that got me is that this is "next" cursor for the page we are on, but even if we stop early, we still persist the page we are on. Thus the nextCursor is always the next cursor.

We should:
- always persist when we have no more entries -- also when there was none at all (this is the problem that started this)
- always persist the initial sync-cursor when doing backfill (not the query cursor one). This will only be needed if we have had 0 results, as if we have had some, then the cursor would already be stored, but if we haven't, then we would never get out of backfill mode. 

Realistic this bcomes "always fires" except when we don't have a cursor. Now, we should always have a cursor, but ... well.. avoiding the match error.

Reply:
Adding couple suggestions to https://github.com/cognitedata/cdp-spark-datasource/pull/917
I think /query case was still missing the save step in some cases, or rather almost all cases as it was only saving if nextCursor existed on last page but it won't be there.

Reply:
I think it covered all cases in the last version, but your suggestion was much more readable and less complex (fewer branches). Pulled in and I think we're good on this now.